### ISSUE TITLE
Error: AnimationController.stop() called after AnimationController.dispose()

### ISSUE THREAD

**Issue by busslina** [GregoryConrad/rearch-dart#212]
I got this error with the first `ToastMessage` the `RearchConsumer` gets (which can change over time if there are multiple messages on a queue).

### Code:
```dart
const _animationDuration = Duration(milliseconds: 100);

class ToastMessageFloatingLayer extends RearchConsumer {
  const ToastMessageFloatingLayer({
    super.key,
    required this.message,
    required this.parentConstraints,
  });

  final ToastMessage message;
  final BoxConstraints parentConstraints;

  @override
  Widget build(BuildContext context, WidgetHandle use) {
    final animationController =
        use.animationController(duration: _animationDuration);

    final animation = use.lazyValue(
      () => CurvedAnimation(
        parent: animationController,
        curve: Curves.fastOutSlowIn,
      ),
    );

    use.effect(
      () {
        animationController.forward();

        return animationController.reset;
      },
      [message],
    );

    return Align(
      alignment: Alignment.bottomCenter,
      child: SizeTransition(
        sizeFactor: animation,
        axis: Axis.vertical,
        child: ToastMessageWidget(
          message: message,
          parentConstraints: parentConstraints,
        ),
      ),
    );
  }
}
```

### Error:
```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following assertion was thrown while finalizing the widget tree:
Assertion failed:
file:///C:/mis-programas/flutter/packages/flutter/lib/src/animation/animation_controller.dart:847:7
_ticker != null
"AnimationController.stop() called after AnimationController.dispose()\nAnimationController methods
should not be used after calling dispose."

When the exception was thrown, this was the stack:
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 296:3     throw_
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 29:3      assertFailed
packages/flutter/src/animation/animation_controller.dart 847:15                 stop
packages/flutter/src/animation/animation_controller.dart 395:5                  set value
packages/flutter/src/animation/animation_controller.dart 416:5                  reset
packages/rearch/src/side_effects.dart 171:67                                    <fn>
packages/flutter_rearch/src/widgets/consumer.dart 124:7                         unmount
packages/flutter/src/widgets/framework.dart 2077:12                             [_unmount]
packages/flutter/src/widgets/framework.dart 2075:7                              <fn>
packages/flutter/src/widgets/framework.dart 6998:9                              visitChildren
packages/flutter/src/widgets/framework.dart 2073:12                             [_unmount]
packages/flutter/src/widgets/framework.dart 2075:7                              <fn>
packages/flutter/src/widgets/framework.dart 5669:7                              visitChildren
packages/flutter/src/widgets/framework.dart 2073:12                             [_unmount]
packages/flutter/src/widgets/framework.dart 2075:7                              <fn>
packages/flutter/src/widgets/layout_builder.dart 123:7                          visitChildren
packages/flutter/src/widgets/framework.dart 2073:12                             [_unmount]
packages/flutter/src/widgets/framework.dart 2075:7                              <fn>
packages/flutter/src/widgets/framework.dart 5669:7                              visitChildren
packages/flutter/src/widgets/framework.dart 2073:12                             [_unmount]
packages/flutter/src/widgets/framework.dart 2075:7                              <fn>
packages/flutter/src/widgets/framework.dart 6998:9                              visitChildren
packages/flutter/src/widgets/framework.dart 2073:12                             [_unmount]
packages/flutter/src/widgets/framework.dart 2075:7                              <fn>
packages/flutter/src/widgets/framework.dart 6886:7                              visitChildren
packages/flutter/src/widgets/framework.dart 2073:12                             [_unmount]
packages/flutter/src/widgets/framework.dart 2075:7                              <fn>
```

**Comment by busslina** [comment]
Seems that the `AnimationController` is already disposed when `effect` calls on `reset()`.

**Comment by busslina** [comment]
Here you are using `memo` with empty dependencies array and then you are disposing on `effect` which triggers on the (wrong?) memoized value.

```dart
AnimationController _animationController(
  WidgetSideEffectRegistrar use, {
  Duration? duration,
  Duration? reverseDuration,
  String? debugLabel,
  double initialValue = 0,
  double lowerBound = 0,
  double upperBound = 1,
  TickerProvider? vsync,
  AnimationBehavior animationBehavior = AnimationBehavior.normal,
}) {
  vsync ??= use.singleTickerProvider();

  final controller = use.memo(
    () => AnimationController(
      vsync: vsync!,
      duration: duration,
      reverseDuration: reverseDuration,
      debugLabel: debugLabel,
      lowerBound: lowerBound,
      upperBound: upperBound,
      animationBehavior: animationBehavior,
      value: initialValue,
    ),
  );
  use.effect(() => controller.dispose, [controller]);

  controller
    ..duration = duration
    ..reverseDuration = reverseDuration;

  return controller;
}
```

**Comment by busslina** [comment]
Maybe I should regenerate the `AnimationController` for each `ToastMessage`, but seems this simple code, reusing it, should work.

**Comment by GregoryConrad** [comment]
> Maybe I should regenerate the AnimationController for each ToastMessage, but seems this simple code, reusing it, should work.


You shouldn't need to. I'll take a look

**Comment by GregoryConrad** [comment]
Not sure there's a bug in ReArch here per se. Just do:

```dart
use.effect(
  () {
    animationController.forward();
    return () {
      if (animationController.isAnimating) animationController.reset();
    };
  },
  [message],
);
```

**Comment by GregoryConrad** [comment]
I'd reckon this is even better:

```dart
use.effect(
  () {
    animationController.forward(from: 0);
    return null;
  },
  [message],
);
```

**Comment by busslina** [comment]
Thanks about your proposed alternatives. But I really believe there is some bug here... So I digged into.

I came to the conclusion that there is a bug in your `_animationController` method. I did some changes in the the `flutter` SDK and in the `flutter_rearch` package in order to see what's happening.

### `animation_controller.dart` (Flutter SDK)
![imagen](https://github.com/user-attachments/assets/c725b8e8-397b-4588-a8d0-4d9e6b3d5607)

### `animation.dart` (flutter_rearch)
```dart
AnimationController _animationController(
  WidgetSideEffectRegistrar use, {
  Duration? duration,
  Duration? reverseDuration,
  String? debugLabel,
  double initialValue = 0,
  double lowerBound = 0,
  double upperBound = 1,
  TickerProvider? vsync,
  AnimationBehavior animationBehavior = AnimationBehavior.normal,
  bool debugMode = false,
}) {
  vsync ??= use.singleTickerProvider();

  if (debugMode) print('_animationController 1');

  final controller = use.memo(
    () => AnimationController(
      vsync: vsync!,
      duration: duration,
      reverseDuration: reverseDuration,
      debugLabel: debugLabel,
      lowerBound: lowerBound,
      upperBound: upperBound,
      animationBehavior: animationBehavior,
      value: initialValue,
    ),
  );
  if (debugMode) print('_animationController 2');

  // use.effect(() => controller.dispose, [controller]);
  use.effect(
      () => () {
            if (debugMode) print('_animationController 3');
            controller.dispose();
            if (debugMode) print('_animationController 4');
          },
      [controller]);

  controller
    ..duration = duration
    ..reverseDuration = reverseDuration;

  return controller;
}
```

### `toast_message_floating_layer.dart` (My project)
```dart
class ToastMessageFloatingLayer extends RearchConsumer {
  const ToastMessageFloatingLayer({
    super.key,
    required this.message,
    required this.parentConstraints,
  });

  final ToastMessage message;
  final BoxConstraints parentConstraints;

  @override
  Widget build(BuildContext context, WidgetHandle use) {
    final animationController =
        use.animationController(duration: _animationDuration, debugMode: true);

    animationController.debugMode = true;

    final animation = use.lazyValue(
      () => CurvedAnimation(
        parent: animationController,
        curve: Curves.fastOutSlowIn,
      ),
    );

    use.effect(
      () {
        print('My project 1');
        animationController.forward();

        // return animationController.reset;
        return () {
          print('My project 2');
          animationController.reset();
          print('My project 3');
        };
      },
      [message],
    );

    use.effect(
      () => () {
        print('My project 4 (Disposing)');
      },
      [],
    );

    return Align(
      alignment: Alignment.bottomCenter,
      child: SizeTransition(
        sizeFactor: animation,
        axis: Axis.vertical,
        child: ToastMessageWidget(
          message: message,
          parentConstraints: parentConstraints,
        ),
      ),
    );
  }
}
```

## Output
```
_animationController 1
_animationController 2
My project 1
_animationController 1
_animationController 2
My project 1
[D]  websocketCapsule -- processMessage() -- <11, _response, bool>
_animationController 3
ANIMATION CONTROLLER DISPOSE
_animationController 4
My project 2
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following assertion was thrown while finalizing the widget tree:
Assertion failed:
file:///C:/mis-programas/flutter/packages/flutter/lib/src/animation/animation_controller.dart:849:7
_ticker != null
"AnimationController.stop() called after AnimationController.dispose()\nAnimationController methods
should not be used after calling dispose."
```

## Conclusion
The `_animationController` function is disposing the `AnimationController with no reason, because the `RearchConsumer` is not disposing.

**Comment by GregoryConrad** [comment]
The issue you’re facing is that disposal order is not currently guaranteed.

If the disposals of use.effects ran in reversed order, you wouldn’t have the issue you’re currently facing (because the reset would be called before the dispose). But right now they aren’t being done in reverse order (I can’t remember if it’s in random order or if it’s in forward order), so you get the assert since the controller is being disposed before your controller.reset call.

now, there is a question of what order disposals/effects should be run in. As mentioned, Right now no guarantees are made since I wanted to have flexibility in internal changes (because of situations like this). a case could be made to make use.effect disposals run in reverse order. That’d also be another way to fix a strange bug I saw previously with animation controllers that I had to add a workaround for (or at least I think reverse order would fix that).

if flutter_hooks or react hooks run effect dispose in reverse order, then I’d be willing to make that change and guarantee it here. But that’d be a separate issue
